### PR TITLE
Fix istioctl upgrade fails for 1.6.x to 1.7.x

### PR DESCRIPTION
**Pull request by carolynhu**
Please provide a description for what this PR is for. 
https://github.com/istio/istio/issues/28595

The current output is:

```
2020-11-29T23:35:47.590678Z     info    Error: failed to generate Istio configs from file [] for the current version: 1.6.8, error: externalComponentSpec for the json field "grafana" is not supported in current version

Error: failed to generate Istio configs from file [] for the current version: 1.6.8, error: externalComponentSpec for the json field "grafana" is not supported in current version
```

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ x] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[x ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any changes that may affect Istio users.

### CODE REVIEW

**Comment by jwendell** [comment]
@carolynhu Could you please provide a release notes entry for this fix?

File: operator/pkg/util/merge_iop.go [hunk 1]
```diff
@@ -275,6 +275,11 @@ func OverlayIOP(base, overlay string) (string, error) {
 
 	merged, err := strategicpatch.StrategicMergePatch(bj, oj, &iopMergeStruct)
 	if err != nil {
+		if strings.Contains(err.Error(), "unable to find api field in struct") {
+			i := strings.Index(err.Error(), "struct")
+			errMsg := err.Error()[i+7:]
+			return "", fmt.Errorf("%s is not supported in current version", errMsg)
```
Comment (shamsher31) [inline operator/pkg/util/merge_iop.go:281]: Is it possible to display the version number?
```suggestion
			return "", fmt.Errorf("%s is not supported from %s version", errMsg, version)
```
```
error: externalComponentSpec for the json field "grafana" is not supported from version 1.7.
```

File: releasenotes/notes/28595.yaml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+apiVersion: release-notes/v2
+kind: bug-fix
```
Comment (jwendell) [inline releasenotes/notes/28595.yaml:2]: Could you link to the bug this fixes?

File: operator/pkg/util/merge_iop.go (hunk 1 above)
Comment (carolynhu) [reply operator/pkg/util/merge_iop.go:281]: yes, it's possible, then i should propagate this error up to the upgrade() func to use the retrieved `targetVersion` there.

File: releasenotes/notes/28595.yaml (hunk 2 above)
Comment (carolynhu) [reply releasenotes/notes/28595.yaml:2]: sure

**Comment by carolynhu** [comment]
The output for this fix is:
```Control Plane - egressgateway pod - istio-egressgateway-787c46c8c4-llt89 - version: 1.6.8
Control Plane - ingressgateway pod - istio-ingressgateway-7bc44b556f-5cbzn - version: 1.6.8
Control Plane - pilot pod - istiod-9b96cbd69-gkk2g - version: 1.6.8

2020-11-30T17:58:34.145671Z     warn    found 4 CRD of unsupported v1alpha1 security policy: [clusterrbacconfigs.rbac.istio.io rbacconfigs.rbac.istio.io servicerolebindings.rbac.istio.io serviceroles.rbac.istio.io]. The v1alpha1 security policy is no longer supported starting 1.6. It's strongly recommended to delete the CRD of the v1alpha1 security policy to avoid applying any of the v1alpha1 security policy in the unsupported version
Upgrade version check passed: 1.6.8 -> 1.7.0.

2020-11-30T17:59:07.668615Z     info    Error: failed to generate Istio configs from file [] for the current version: 1.6.8, error: externalComponentSpec for the json field "grafana" is not supported in 1.7.0

Error: failed to generate Istio configs from file [] for the current version: 1.6.8, error: externalComponentSpec for the json field "grafana" is not supported in 1.7.0
```

**Review by ostromart — CHANGES_REQUESTED** [review]
this is not the right approach. we must find where the new schema is used to unmarshal old data and fix the issue there.

**Comment by carolynhu** [comment]
cc @elfinhe

**Comment by elfinhe** [comment]
Another option - when reading profile from a file path, we check addon components, give warnings, and then remove addon components if they are from previous built-in profiles.

https://github.com/istio/istio/blob/7a7a0d39ca4a97a94cf5f83f8bac5a923a51736f/operator/pkg/helm/helm.go#L91

**Comment by jwendell** [comment]
Folks, is this still relevant or safe to close?

**Comment by jwendell** [comment]
Are you going to provide another PR?

**Comment by jwendell** [comment]
Closing, feel free to reopen if you think this is a mistake.

**Pull request closed without merging**